### DISCUSSION TITLE
Резервная копия аккаунта

### DISCUSSION THREAD

**Discussion by amd64fox** [Chimildic/goofy#170] [Личный опыт]
В преддверии бана своего основного аккаунта, вдруг захотел реализовать резерв всего чего можно.

### Что делает код:
Создает резервную копию на гугл диск и восстанавливает следующие элементы:
- Любимые треки
- Отслеживаемые артисты
- Отслеживаемые альбомы
- Отслеживаемые плейлисты
- Личные плейлисты

Учитывайте пожалуйста, что все зарезервированные плейлисты которые находились у вас в папках будут восстановлены без них, так как API Spotify не поддерживает создание папок.

Есть также один недочет, сортировка от модуля [Library](https://chimildic.github.io/goofy/#/reference/library?id=library) хаотичная, а это значит, что при восстановлении у любимых треков, отслеживаемых артистов и альбомов будет нарушена оригинальная сортировка, как это исправить пока не знаю, если есть идеи пишите ниже.

### Как пользоваться:
Использовать максимально просто, если нужно сделать резервную копию раскомментируйте функцию `create()` и наоборот если нужно восстановить сделанную копию закомментируйте обратно `create()` и раскомментируйте `restore()`
```javascript
function backup() {

    // create()  // Создать резервную копию 
    // restore() // Восстановить резервную копию

    function create() {

        // Резерв любимых треков
        let love_tracks = Source.getSavedTracks();
        Cache.compressTracks(love_tracks);
        Cache.write('root/backup/love_tracks.json', love_tracks);

        // Резерв отслеживаемых артистов
        let followArtists = Source.getArtists({ followed_include: true });
        Cache.write('root/backup/follow_artists.json', followArtists);

        // Резерв отслеживаемых альбомов
        let followAlbums = Source.getSavedAlbums();
        Cache.write('root/backup/follow_albums.json', followAlbums)

        // Резерв отслеживаемых плейлистов
        function followedPlaylists(type = 'followed', userId = User.id) {
            let playlistArray = Playlist.getPlaylistArray(userId);
            playlistArray = playlistArray.filter((playlist) => {
                let isOwned = playlist.owner.id == userId;
                return type == 'owned' ? isOwned : !isOwned;
            });
            return playlistArray;
        }
        Cache.write('root/backup/follow_playlists.json', followedPlaylists())

        // Резерв личных плейлистов
        let playlists = Playlist.getPlaylistArray(User.id).filter(item => item.owner.id == User.id)
        let tracks = Source.getTracks(playlists);
        Cache.compressTracks(tracks);
        playlists.forEach(p => (p.tracks = tracks.filter(t => t.origin.id == p.id)));
        Order.reverse(playlists);
        Cache.write('root/backup/owned_playlists.json', playlists);

    }

    function restore() {

        // Восстановление любимых треков
        let love_tracks = Cache.read('root/backup/love_tracks.json');
        Library.saveFavoriteTracks(love_tracks);

        // Восстановление отслеживаемых артистов
        let followArtists = Cache.read('root/backup/follow_artists.json');
        Library.followArtists(followArtists);

        // Восстановление отслеживаемых альбомов
        let followAlbums = Cache.read('root/backup/follow_albums.json');
        Library.saveAlbums(followAlbums);

        // Восстановление отслеживаемых плейлистов
        let followPlaylists = Cache.read('root/backup/follow_playlists.json');
        Library.followPlaylists(followPlaylists);


        // Восстановление личных плейлистов
        let playlists = Cache.read('root/backup/owned_playlists.json');
        playlists.forEach(p => {
            let params = {
                name: p.name,
                tracks: p.tracks,
                description: p.description,
                public: p.public,
            }
            p.images.length > 0 && (params.sourceCover = p.images[0].url);
            Playlist.saveAsNew(params)
        });

    }
}
```
И так, с активированной функцией `create()` раз в неделю например будет перезаписываться новая копия любимых треков, отслеживаемых артистов, альбомов и плейлистов.

В библиотеке как я понял не было опубликованной функции, что бы сразу получить все отслеживаемые плейлисты, проблема решилась когда я нашел неопубликованную функцию которая ссылалась на `getFollowedTracks` в библиотеке, чуть-чуть сократил и вставил в свой код, получилось то что нужно.

Но вот так же сделать с личными плейлистами, не выходит. Переключил на `owned`, но в массив заливается инфа о моих плейлистах и все.

Есть какой-то просто способ сохранить все свои личные плейлисты в кеш ? И потом достать все это в том же виде, но в другом аккаунте ?

Проблема еще в том что, в личных плейлистах много треков, возможно есть риски не вписаться в лимиты.

**Comment by Chimildic** [comment]
`Playlist.getPlaylistArray` дает данные о плейлисте без треков. Их нужно запрашивать дополнительно. Так устроено API Spotify.

> в массив заливается инфа о моих плейлистах и все

Вы хотите сохранить _id_ плейлистов, чтобы подписаться на них в новом аккаунте или сохранить треки в кэш и потом создать копию плейлиста?

В первом случае, ваш вариант `followedPlaylists` с `owned` даст эти _id_.

**Reply by amd64fox** [reply]
Да, я хочу создать полную копию всех личных плейлистов.

Если предположить что старый аккаунт будет удален то вариант передачи id плейлиста через `owned` не подходит т.к. его уже по сути не будет существовать.

**Reply by Chimildic** [reply]
Создание резервной копии
```js
  let playlists = Playlist.getPlaylistArray(User.id).filter(item => item.owner.id == User.id)
  let tracks = Source.getTracks(playlists);
  Cache.compressTracks(tracks);
  playlists.forEach(p => (p.tracks = tracks.filter(t => t.origin.id == p.id)));
  Cache.write('root/backup/owned_playlists.json', playlists);
```

Восстановление из резервной копии
```js
  let playlists = Cache.read('root/backup/owned_playlists.json');
  playlists.forEach(p => {
    let params = {
      name: p.name,
      tracks: p.tracks,
      description: p.description,
      public: p.public,
    }
    p.images.length > 0 && (params.sourceCover = p.images[0].url);
    Playlist.saveAsNew(params)
  });
```

**Reply by amd64fox** [reply]
Ох.. благодарю за готовое решение. 
Сейчас потестил, создание копии отрабатывает четко и быстро, а восстановление прерываться где-то в конце с ошибкой:

```
TypeError: Cannot read property 'url' of undefined
(функция без названия)	@ main.gs:11
test	@ main.gs:5
```
```javascript
function test() {


  let playlists = Cache.read('root/backup/owned_playlists.json');
  playlists.forEach(p => {
    Playlist.saveAsNew({
      name: p.name,
      tracks: p.tracks,
      description: p.description,
      public: p.public,
      sourceCover: p.images[0].url
    })
  });

console.log('Число запросов', CustomUrlFetchApp.getCountRequest());

}
```
Если правильно понял, ругается на url какой-то картинки.

**Reply by amd64fox** [reply]
Залез сейчас в массив который был создан, просмотрел бегло и нашел тот самый `images` на который вылетает ошибка.
<details>
<summary><small>Скрин</small></summary><p>

![Снимок экрана 2022-02-13 062201](https://user-images.githubusercontent.com/62529699/153740128-a3e6e43c-4de7-4ad6-aca3-9f654d83b00e.jpg)

</details>

Этот плейлист единственный у которого нет треков и нет кастомного изображения, получается в `"images":[]`  и ошибка с прерыванием кода

Повторно проверил без этого плейлиста, код выполняется без ошибок, это возможно исправить ?

**Reply by Chimildic** [reply]
Обновил код выше

**Reply by amd64fox** [reply]
Спасибо, работает.
Единственно заметил еще, что все `Library.` почему-то восстанавливают хаотичную сортировку, но это в принципе мелочь.

**Comment by amd64fox** [comment]
В общем, сделал тестовый бекап и восстановление на другом аккаунте.

Мой аккаунт содержал на момент бекапа :

Подписанные артисты: 273
Подписанные альбомы: 49
Подписанные плейлисты: 23 (2807 треков)
Любимые треки: 3254 треков
Мои плейлисты: 67 (34694 треков)

`Создание резерва:`
Время: 55c.
Запросов: 764

<details>
<summary><small>файлы на диске</small></summary><p>

![Снимок экрана 2022-02-13 112143](https://user-images.githubusercontent.com/62529699/153745062-0532494e-91d1-4dbf-aa33-6f7435bb61c0.jpg)


</details>

`Восстановление:`
Время: 240с.
Запросов: 668
